packages/firestore/src/model/proto_values.ts

```diff
@@ -0,0 +1,385 @@
+    return TypeOrder.StringValue;
+  } else if ('bytesValue' in value) {
+    return TypeOrder.BlobValue;
+  } else if ('referenceValue' in value) {
+    return TypeOrder.RefValue;
+  } else if ('geoPointValue' in value) {
+    return TypeOrder.GeoPointValue;
+  } else if ('arrayValue' in value) {
+    return TypeOrder.ArrayValue;
+  } else if ('mapValue' in value) {
+    return TypeOrder.ObjectValue;
+  } else {
+    return fail('Invalid value type: ' + JSON.stringify(value));
+  }
+}
+
+/** Returns whether `value` is defined and corresponds to the given type order. */
+export function isType(
+  value: api.Value | undefined,
+  expectedTypeOrder: TypeOrder
+): boolean {
+  return value !== undefined && typeOrder(value) === expectedTypeOrder;
+}
+
+/** Tests `left` and `right` for equality based on the backend semantics. */
+export function equals(left: api.Value, right: api.Value): boolean {
+  const leftType = typeOrder(left);
+  const rightType = typeOrder(right);
+  if (leftType !== rightType) {
+    return false;
+  }
+
+  switch (leftType) {
+    case TypeOrder.NullValue:
+      return true;
+    case TypeOrder.BooleanValue:
+      return left.booleanValue === right.booleanValue;
+    case TypeOrder.TimestampValue:
+      return timestampEquals(left, right);
+    case TypeOrder.StringValue:
+      return left.stringValue === right.stringValue;
+    case TypeOrder.BlobValue:
+      return blobEquals(left, right);
+    case TypeOrder.RefValue:
+      return left.referenceValue === right.referenceValue;
+    case TypeOrder.GeoPointValue:
+      return geoPointEquals(left, right);
+    case TypeOrder.NumberValue:
+      return numberEquals(left, right);
+    case TypeOrder.ArrayValue:
+      return arrayEquals(left, right);
+    case TypeOrder.ObjectValue:
+      return objectEquals(left, right);
+    default:
+      return fail('Unexpected value type: ' + JSON.stringify(left));
+  }
+}
+
+function timestampEquals(left: api.Value, right: api.Value): boolean {
+  const leftTimestamp = normalizeTimestamp(left.timestampValue!);
+  const rightTimestamp = normalizeTimestamp(right.timestampValue!);
+  return (
+    leftTimestamp.seconds === rightTimestamp.seconds &&
+    leftTimestamp.nanos === rightTimestamp.nanos
+  );
+}
+
+function geoPointEquals(left: api.Value, right: api.Value): boolean {
+  return (
+    normalizeNumber(left.geoPointValue!.latitude) ===
+      normalizeNumber(right.geoPointValue!.latitude) &&
+    normalizeNumber(left.geoPointValue!.longitude) ===
+      normalizeNumber(right.geoPointValue!.longitude)
+  );
+}
+
+function blobEquals(left: api.Value, right: api.Value): boolean {
+  return normalizeByteString(left.bytesValue!).isEqual(
+    normalizeByteString(right.bytesValue!)
+  );
+}
+
+export function numberEquals(left: api.Value, right: api.Value): boolean {
+  if ('integerValue' in left && 'integerValue' in right) {
+    return (
+      normalizeNumber(left.integerValue) === normalizeNumber(right.integerValue)
```

Comment:
Should the serializer just perform these normalizations once so we don't have to do it on every comparison?